src/models/models.jl

```diff
@@ -110,22 +116,112 @@ function InfiniteMPOMatrix(model::Model, s::CelledVector, translator::Function;
       if isnothing(idx)
         Hmat[range_H + 1 - n, range_H - n] = identity
       else
-        Hmat[range_H + 1 - n, range_H - n] = temp_H[j - n][idx]#replacetags(linkinds, temp_H[j - n][idx], "Link, l=$n", tags(ls[j-1]))
+        #Here, we split the local tensor into its different blocks
+        T = eltype(temp_H[j - n][idx])
+        temp_mat = convert_itensor_to_itensormatrix(
+          temp_H[j - n][idx];
+          leftdir=ITensors.In,
+          first=n == 0 ? true : false,
+          last=n == range_H - 1 ? true : false,
+          left_tags=tags(ls[j - 1]),
+          right_tags=tags(ls[j]),
+          leftindex=if idx > 1
+            only(commoninds(temp_H[j - n][idx], temp_H[j - n][idx - 1]))
+          else
+            nothing
+          end,
+        )
+        if size(temp_mat) == (3, 3)
+          @assert iszero(temp_mat[1, 2])
+          @assert iszero(temp_mat[1, 3])
+          @assert iszero(temp_mat[2, 3])
+          @assert temp_mat[1, 1] == identity
+          @assert temp_mat[3, 3] == identity
+          Hmat[range_H + 1 - n, range_H - n] = temp_mat[2, 2]
+          Hmat[end, range_H - n] = temp_mat[3, 2]
+          Hmat[range_H + 1 - n, 1] = temp_mat[2, 1]
+        elseif size(temp_mat) == (1, 3)
+          @assert n == 0
+          @assert temp_mat[1, 3] == identity
+          #@assert isempty(temp_mat[1, 1]) || iszero(temp_mat[1, 1])
+          Hmat[range_H + 1 - n, range_H - n] = temp_mat[1, 2]
+          Hmat[range_H + 1 - n, 1] = temp_mat[1, 1]
+        elseif size(temp_mat) == (3, 1)
+          @assert (range_H - n) == 1
+          @assert temp_mat[1, 1] == identity
+          #@assert isempty(temp_mat[3, 1]) || iszero(temp_mat[3, 1])
+          Hmat[range_H + 1 - n, range_H - n] = temp_mat[2, 1]
+          Hmat[end, range_H - n] += temp_mat[3, 1]  #LH This should do nothing #TODO check
+        else
+          error("Unexpected matrix form")
+        end
       end
     end
     mpos[j] = Hmat
     #mpos[j] += dense(Hmat) * setelt(ls[j-1] => total_dim) * setelt(ls[j] => total_dim)
   end
-  #return mpos
-  return InfiniteMPOMatrix(mpos, translator)
+  #unify_indices and add virtual indices to the empty tensors
+  mpos = InfiniteMPOMatrix(mpos, translator)
+  for x in 1:N
+    sp = prime(s[x])
+    sd = dag(s[x])
+    left_inds = [
+      only(uniqueinds(mpos.data.data[x][j, 1], mpos.data.data[x][1, 1])) for
```

Comment:
Reaching into the data structures like this is definitely not recommended/encouraged, is there a better way to do this?

Reply:
So two things here
- I noticed a significant slow down/cost to do the operations at the mpos level, weirdly. Again the translatecell calls seem to have a small effect on performances (mainly because we are calling this _a_lot_, Normally the new functions should be better, though it is still visible.
I nonetheless removed all the direct accesses that are not necessary (see below)
- For some operations, it is required as we are modifying the tensors. Modifying the tensors at the global MPO level did not seem to work.